### PR TITLE
[XCConfig Helper] Add dependency target vendored libs and frameworks to build settings.

### PR DESCRIPTION
**Pull request by kcoleman731**
🌈 :thumbsup: 

This pull requests adds support for recursively processing `dependent_targets` to ensure that all `vendored_frameworks` and `vendored_libraries` are included in the target's build settings. This is only done if `use_frameworks` is specified.

### CODE REVIEW

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 1]
```diff
@@ -91,7 +99,7 @@ def self.add_spec_build_settings_to_xcconfig(consumer, xcconfig)
         #
         def self.add_framework_build_settings(framework_path, xcconfig, sandbox_root)
           name = File.basename(framework_path, '.framework')
-          dirname = '$(PODS_ROOT)/' + framework_path.dirname.relative_path_from(sandbox_root).to_s
+          dirname = '${PODS_ROOT}/' + framework_path.dirname.relative_path_from(sandbox_root).to_s
```
Comment (segiddins) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: why these changes?

**Comment by segiddins** [comment]
https://travis-ci.org/CocoaPods/CocoaPods/jobs/82608982#L2475 looks incorrect to me? (shouldn't it only be adding frameworks of transitive dependencies, not libraries?)

**Comment by segiddins** [comment]
Also needs a CHANGELOG entry, and review from @mrackwitz

**Comment by kcoleman731** [comment]
Thanks for the feedback @segiddins...looking into this now.

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb (hunk 1 above)
Comment (kcoleman731) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: When I ran the `install_vendored_dynamic_framework` integration spec, it initially spit out the following 
```
FRAMEWORK_SEARCH_PATHS = FRAMEWORK_SEARCH_PATHS = $(inherited) "${PODS_ROOT}/Realm/realm-swift-0.94.1/ios/swift-1.2" "$(PODS_ROOT)/Realm/realm-swift-0.94.1/ios/swift-1.2" 
```

I took care of the duplication, but figured we wanted to go with curly braces (since the duplication my doing). Happy to change it back.

**Comment by kcoleman731** [comment]
@segiddins have a look at the integration spec now. You issue is addressed. Thanks for calling that out.

**Comment by segiddins** [comment]
:+1: @mrackwitz please also review

**Comment by blakewatters** [comment]
@segiddins @mrackwitz Anything we can do to help move this along? We are trying to do some fun release choreography across the dynamic framework and pod. Willing to do whatever leg work is necessary

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 2]
```diff
@@ -50,6 +50,14 @@ def self.default_ld_flags(target, includes_static_libraries = false)
         #         The xcconfig to edit.
         #
         def self.add_settings_for_file_accessors_of_target(target, xcconfig)
+          # Recursively process all dependent targets to ensure dependency vendored
+          # frameworks and libraries are added to OTHER_LDFLAGS and FRAMEWORK_SEARCH_PATHS.
+          if target.host_requires_frameworks?
```
Comment (mrackwitz) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: I think this should be `requires_frameworks?`

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 3]
```diff
@@ -52,6 +52,26 @@ def self.default_ld_flags(target, includes_static_libraries = false)
...
+        # Adds build settings for vendored frameworks and libraries.
+        #
+        # @param [PodTarget] target
+        #        The pod target, which holds the list of +Spec::FileAccessor+.
+        #
+        # @param [Xcodeproj::Config] xcconfig
+        #        The xcconfig to edit.
+        #
+        def self.add_vendored_dependency_build_settings(target, xcconfig)
+          if target.host_requires_frameworks?
```
Comment (mrackwitz) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: I think this should be `requires_frameworks?`

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 4]
```diff
@@ -52,6 +52,26 @@ def self.default_ld_flags(target, includes_static_libraries = false)
...
+            target.dependent_targets.each do |dependent_target|
+              XCConfigHelper.add_vendored_dependency_build_settings(dependent_target, xcconfig)
```
Comment (mrackwitz) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: I don't think we need to recurse here. Dependent targets includes already transitive dependencies.

File: CHANGELOG.md [hunk 5]
```diff
@@ -23,6 +23,9 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
   [Samuel Giddins](https://github.com/segiddins)
   [#4124](https://github.com/CocoaPods/CocoaPods/issues/4124)
 
+* Support for adding dependency target vendored libraries and frameworks to build settings.
```
Comment (mrackwitz) [inline CHANGELOG.md]: * [ ] Add two spaces at the end of the description to force a line-mark.
* [ ] Add a link to this PR for reference below your name.

**Comment by mrackwitz** [comment]
@kcoleman731 @blakewatters: It's unclear to me why this is needed at all. Can you explain it for the example, where you noticed it yourself?

**Comment by blakewatters** [comment]
Our vendored framework was not getting pulled in and we were hitting build errors. @kcoleman731 can get you a pod spec and a framework (built from cocoapods-packager) that fails to integrate cleanly without the patch

**Comment by segiddins** [comment]
@mrackwitz it's needed if a source pod depends on a dynamic vendored framework pod. I confirmed that adding the framework linker flag is necessary.

**Comment by kcoleman731** [comment]
@mrackwitz @segiddins 
- [x] Change `target.host_requires_frameworks?` to `target.requires_frameworks?`
- [x] Add two spaces at the end of the description to force a line-mark.
- [x] Add a link to this PR for reference below your name.

Let me know if there is anything else.

**Comment by segiddins** [comment]
:ship: thanks so much @kcoleman731
